woocommerce/woocommerce
# Remove shipping if there is no address, always allow hide shipping costs option

Closes #52335 

With some adjustments:

1. We used to disable "hide shipping costs until an address is entered" when local pickup was enabled. We now allow it to be enabled but we explain the behaviour (`Local pickup rates will display in the Checkout block, even without an address`). @opr and I agreed this experience felt clearer and covered all cases well, but it would be good to get some feedback about this cc @pmcpinto 

2. If an address has not been entered and the user has hide shipping costs enabled we do not display the *Shipping* options

Added by @opr 

- Changes the shipping calculator logic to display "Enter an address to check delivery options" if only a partial address has been entered (e.g. when the shopper has a default location)
- We previously hid the "Ship" selector when:
  - Shipping is not required
  - Shipping calculations have not been completed
  - No shipping rates are available
  - The item is not collectible
  - Local pickup is not enabled
  - No shipping methods are currently available
- however this would hide it even if rates _might_ become available later, after a full address had been entered. Instead I changed the logic to force the control to show if the following is true:
  - Shipping is enabled, and
  - The cart needs shipping, and
  - All packages in the cart support collectible, and
  - Local pickup is enabled, and
  - Shipping methods exist on the site
- Updated shipping calculator so it doesn't show a "No delivery options available for XXX, XX" message unless there's a full address. I figured the shopper could be negatively affected by such a message if they haven't even filled their address in and they're already seeing a message saying there are no shipping methods. Because of this, I updated the tests in `plugins/woocommerce-blocks/tests/e2e/tests/cart/cart-checkout-block-shipping.block_theme.spec.ts`.
- Added a `has_full_shipping_address` method to `ShippingController` - this uses `WC()->countries->get_country_locale` to determine whether the state, city, country, postcode (based on locale requirement) is filled before stripping out "shipping" methods (and leaving pickup methods intact).
  - Needed because we have to determine whether to remove rates from the response if shipping requires a full address.
  

### How to test the changes in this Pull Request:

Smoke test a few scenarios to ensure no regressions. For the shopper part it is best to open an incognito window to clear the address state that would be remembered between scenarios.

#### Scenario 1
1. Have local pickup disabled
2. Have hide shipping costs until address entered **enabled**
3. As shopper go through cart and checkout and ensure that shipping costs are not shown on cart page / checkout page until address entered

#### Scenario 2

1.  Have local pickup disabled
2. Have hide shipping costs until address entered **disabled**
3. As a shopper go through cart and checkout and ensure that shipping and pick up options are visible before address is entered

#### Scenario 3

1. Have local pickup enabled
2. Have hide shipping costs until address entered **enabled**
3.  As a shopper go through cart and checkout and ensure that shipping options are not shown but pickup options are still shown.

#### Scenario 4

1. Go to WC -> Settings -> General. Set Shipping location(s) to `Disable shipping & shipping calculations`
2. Try to check out with a physical product. You should not be able to enter a shipping address or select local pickup. Checkout should succeed.
3. Try to check out with a virtual product. You should not be able to enter a shipping address or select local pickup. Checkout should succeed.

#### General testing

Test with various setups of:
- Local pickup enabled/disabled
- Hide shipping costs until an address is entered enabled/disabled
- Shipping calculator enabled/disabled
- No shipping zones enabled on the whole site
- Only fallback shipping zones enabled
- No fallback zones enabled, but methods exist for specific countries. Test with addresses in and out of these countries.
- Mixtures of virtual, downloadable, and physical products in the cart. Try with each type individually too.

### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 
Improve how we show/hide shipping address at cart and checkout based on options.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
